allow reading from tarball on stdin

josch opened the pull request:
This pull request fixes https://github.com/bestouff/genext2fs/issues/10

The user interface is changed like this: when no input is given via -d or -D, then genext2fs will open standard input and try reading a tarball from it.

By default, genext2fs tries to create the minimum number of inodes by reading the input twice. This is of course not possible when the input is read from stdin. Thus, the user has to supply the number of inodes manually. To not make this too painful, the new special value 0 got introduced for the -N option which means: compute inode number automatically from the blocksize and number of blocks.

With this commit I can successfully create a full Debian system image like this:

    ./genext2fs -b 144000 -N 0 disk.img < chroot.tar

This is a work in progress. What do you think?

bestouff commented:
Could you please rebase to latest master, and check tests pass ?

josch commented:
Done. It seems the tests are also run by your automatic testsuite that gets run for pull requests, so you don't even have to take my word for it. :)

I'm just still wondering whether you like the inferface? Currently, one has to add `-N 0` to automatically compute a good number of inodes. Maybe you would like something else? For example if the input is a tarball, then it might make sense to automatically auto-compute the inode number without having to explicitly set `-N 0`?

bestouff commented:
I'm not a fan of the interface, no. And if , as I suggested, you add a specific option for archives (e.g. `-a`) then `-N 0` isn't needed.  
You can perhaps re-add the `-N 0` feature but as a separate merge request, with helptext and tests update if possible.

bestouff commented:
> This is a work in progress. What do you think?

To make it clear: this is a great feature and I like it, I just don't want to reserve stdin for that.  
Also would it be possible for you to add tests (e.g. a small tarfile and associated `test.sh` fingerprint) ?

bestouff commented on genext2fs.c:
```diff
@@ -3362,16 +3486,20 @@ main(int argc, char **argv)
 		stats.nblocks = 0;
 
-		populate_fs(NULL, dopt, didx, squash_uids, squash_perms, fs_timestamp, &stats);
+		if (nbinodes == 0) {
```
This feature should be in its own merge request.

josch replied:
I don't disagree. But what other way do you see to compute the number of inodes for when the tarball comes from stdin? In that case, we cannot call `populate_fs` twice to figure out the required number of inodes in the first call. One way would be to require the `-i` option (bytes per inode) being set to auto-compute the value from the given number of blocks.

bestouff replied:
`-t -` (or even `-D -`) would imply a mode without precomputing inodes number, and should require `-N 0` or something else specifying this number. But as I said, it's another feature for another merge request.

josch commented on genext2fs.c:
```diff
@@ -2278,6 +2297,101 @@ add2fs_from_file(filesystem *fs, uint32 this_nod, FILE * fh, uint32 fs_timestamp
+	if (archive_read_support_format_all(a) != ARCHIVE_OK)
+		error_msg_and_die("Couldn't enable read formats");
+
+	if ((r = archive_read_open_fd(a, 0, 10240)))
```
Thank you for your input! Okay, so you'd rather add a new option like `-t` and the argument for that option would then be the tarball? Then using `-t -` could be made to mean "read from standard input" as the `-` is also used to mean "standard input" in other tools. Would that be okay?

bestouff replied:
That would be okay, yes.

bestouff replied:
Thinking a bit more about it, it would be high time to really separate `-d` and `-D`. So `dopt` shouldn't be a `char * dopt[MAX_DOPT]` but instead something like a `struct fspiece dopt[MAX_DOPT]` where
```c
struct fspiece
{
    char piece_type;
    char *piece_path;
}
```
Directories would have type `'d'`, device tables type `'D'`, tarballs type `'t'`. Or maybe an enum would be cleaner.
Would that be ok with you ?

bestouff replied:
One more thing, I realized `-t` is already used by `mke2fs` with which I try to (very loosely) maintain options compatibility. So better use `-a`.

bestouff replied:
I have done that part: there's now a `fslayer` struct (with already a FSLAYER_TAR type for you) which you should use for calling `add2fs_from_tar()` in `populate_fs()`. You should rebase.

josch commented on configure.in:
```diff
@@ -36,6 +36,28 @@ AC_CHECK_FUNCS([getopt_long getline strtof])
 AC_FUNC_SNPRINTF
 AC_FUNC_SCANF_CAN_MALLOC
 
+AC_MSG_CHECKING(--enable-libarchive argument)
```
Okay, I will change the default and instead add a `--disable-libarchive` option to `./configure`. My initial assumption was, that you'd not like the additional dependency because libarchive is quite heavy drawing in all kinds of compression libraries and even libxml, which is quite a drastic change for a tool like genext2fs which didn't have any library dependencies before.

bestouff replied:
Oh effectively, that looks terrible.
I know I already wrote some tar parsing code once, it's not that complex. Maybe one day I'll do it again :)

bestouff commented:
@josh I have tried to implement your idea without external dependencies (yes I know, I have the NIH syndrome). Could you have a look at https://github.com/bestouff/genext2fs/tree/xav-tarball and tell me what you think ? It's not finished yet, not tested, but seems to work on simple tarballs.
I hope you don't mind.

josch commented:
> Also would it be possible for you to add tests (e.g. a small tarfile and associated `test.sh` fingerprint) ?

I can add a small tarfile but maybe such a "binary blob" would not be elegant? Other test input is also generated on the fly. I can easily generate a tarball but would you permit doing so using Perl or Python (either without needing additional modules). Generating binary data in shell is really painful and I'd like to avoid using the `tar` command because its output depends on the implementation (GNU tar versus BSD tar) as well as on the version.

So would you prefer Perl or Python script (any preference? I already have code in Perl) over just dropping a small tarball into the git repository? Alternatively, I can also produce a tarball via a C program -- still much better than using sh...

josch commented:
> @josh I have tried to implement your idea without external dependencies (yes I know, I have the NIH syndrome). Could you have a look at https://github.com/bestouff/genext2fs/tree/xav-tarball and tell me what you think ? It's not finished yet, not tested, but seems to work on simple tarballs.
> I hope you don't mind.

Cool! I'd still like the option to use libarchive though, because I doubt that you plan to support all kinds of compression algorithms, ustar, posix (pax), oldgnu, sparse files, extended attributes, extended headers, user/group name resolution, long names, multivolume offsets, ACLs, and so on...

If you like, I can rebase my work on top of yours, such that the `--with-libarchive` switch will instruct genext2fs to use libarchive instead of your code. I can understand the desire to have no dependencies but "tar" is a really complicated and overloaded format which needs thousands of lines of code to support all the weird corner cases (and even then you'll still have bugs). So as the Debian maintainer of genext2fs I rather would build genext2fs with libarchive.

bestouff commented:
> f you like, I can rebase my work on top of yours, such that the --with-libarchive switch will instruct genext2fs to use libarchive instead of your code.

Ok for that.

josch commented:
Cool! :)

Then just ping this pull request once you put your work in the master branch and then I'll rebase. Thanks!

bestouff commented:
> I can add a small tarfile but maybe such a "binary blob" would not be elegant?

Not in the `git` spirit, I know. But that's the easiest and more robust solution I can think of.

bestouff commented:
Here we are - sorry for the delay.  
@josch could you rebase ?

josch commented:
I rebased my work on top of the current master branch. This only adds support for reading tarballs with libarchive and does not touch command line arguments or other behaviour. The `--enable-libarchive` option is added to `configure` so that the feature will only be enabled if requested. To this end I added a big `#ifdef` into the body of `add2fs_from_tarball` which will pick the right parser depending on whether libarchive was enabled during configure or not.

One caveat: you modified `mkfile_fs()` to receive a parameter `size`. Some of the archive formats supported by libarchive do not carry the entry file size so with libarchive, the size will sometimes be zero. This does not play well with how `mkfile_fs()` operates, so I split out, the reading loop into one version for plain `FILE *` handles and one for `struct archive *` types. For the latter we don't need to know the total file size, so the parameter is ignored in that case. This property also means that the inode counting will be off in cases where the archive does not contain the entry file size. I added a code comment about that.

Lastly, I also added a new test. I created a tarball containing a regular file, a directory, a fifo and a character device and compressed the result with gzip. Since the gzipped tarball is very tiny, I embedded it base64 encoded into `test.sh`. Notice, that `test.sh` currently fails, because it is run without libarchive and your code currently ignores permissions for group and others in all cases but for symlinks. Once you fix that, the test case should also pass without libarchive.

josch commented:
Turns out, your code was fine and there was a problem in mine with the `-q` option. I think you can merge now. Test cases pass with and without libarchive.

bestouff commented:
Thank you I started to be a bit lost in the tar formats :)
